alerting/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/util/AnomalyDetectionUtils.kt

```diff
@@ -0,0 +1,77 @@
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.alerting.util
+
+import com.amazon.opendistroforelasticsearch.alerting.core.model.SearchInput
+import com.amazon.opendistroforelasticsearch.alerting.model.Monitor
+import com.amazon.opendistroforelasticsearch.commons.authuser.User
+import org.apache.lucene.search.join.ScoreMode
+import org.elasticsearch.index.query.BoolQueryBuilder
+import org.elasticsearch.index.query.NestedQueryBuilder
+import org.elasticsearch.index.query.QueryBuilders
+import org.elasticsearch.search.builder.SearchSourceBuilder
+
+/**
+ * AD monitor is search input monitor on top of anomaly result index. This method will return
+ * true if monitor input only contains anomaly result index.
+ */
+fun isADMonitor(monitor: Monitor): Boolean {
+    // If monitor has other input than AD result index, it's not AD monitor
+    if (monitor.inputs.size != 1) {
+        return false
+    }
+    val input = monitor.inputs[0]
+    // AD monitor can only have 1 anomaly result index.
+    if (input is SearchInput && input.indices.size == 1 && input.indices[0] == ".opendistro-anomaly-results*") {
+        return true
+    }
+    return false
+}
+
+fun addUserBackendRolesFilter(user: User?, searchSourceBuilder: SearchSourceBuilder): SearchSourceBuilder {
+    var boolQueryBuilder = BoolQueryBuilder()
+    val userFieldName = "user"
+    val userBackendRoleFieldName = "user.backend_roles.keyword"
+    if (user == null) {
+        // For old monitor and detector, they have no user field
+        val userRolesFilterQuery = QueryBuilders.existsQuery(userFieldName)
+        val nestedQueryBuilder = NestedQueryBuilder(userFieldName, userRolesFilterQuery, ScoreMode.None)
+        boolQueryBuilder.mustNot(nestedQueryBuilder)
+    } else if (user.backendRoles.isNullOrEmpty()) {
+        // For simple FGAC user, they may have no backend roles, these users should be able to see detectors
+        // of other users whose backend role is empty.
+        val userRolesFilterQuery = QueryBuilders.existsQuery(userBackendRoleFieldName)
+        val nestedQueryBuilder = NestedQueryBuilder(userFieldName, userRolesFilterQuery, ScoreMode.None)
+
+        val userExistsQuery = QueryBuilders.existsQuery(userFieldName)
+        val userExistsNestedQueryBuilder = NestedQueryBuilder(userFieldName, userExistsQuery, ScoreMode.None)
+
+        boolQueryBuilder.mustNot(nestedQueryBuilder)
+        boolQueryBuilder.must(userExistsNestedQueryBuilder)
+    } else {
+        // For normal case, user should have backend roles.
+        val userRolesFilterQuery = QueryBuilders.termsQuery(userBackendRoleFieldName, user.backendRoles)
+        val nestedQueryBuilder = NestedQueryBuilder(userFieldName, userRolesFilterQuery, ScoreMode.None)
+        boolQueryBuilder.must(nestedQueryBuilder)
```

Comment:
These 2 lines are repeated in all of the different `if else` cases

Reply:
Not all branches. line 68 and line 63 looks like duplicate, but if I remove one line of them, I need to add another 2 lines: 
Sample code
```
var mustQuery = null
if (...) {
   mustQuery = userExistsNestedQueryBuilder
} else {
   mustQuery = nestedQueryBuilder
}
boolQueryBuilder.must(mustQuery)
```
Prefer to keep as is.

Reply:
ah ok, didnt realize that. Thanks